### ISSUE TITLE
GraphSaint arch: '0-0-0-0' vs MLP sanity check

### ISSUE THREAD

**Issue by omgmax** [GraphSAINT/GraphSAINT#27]
Hi there, 

If my understanding is correct, setting `arch: '0-0-0-0'` in the `train_config` would be equivalent to an MLP. 

I trained a GraphSAINT model with this architecture and compared it to an MLP as a sanity check and got much worse performance for the GraphSAINT model:
![image](https://user-images.githubusercontent.com/73133329/98833798-7705ac00-240c-11eb-9da2-ec0705319863.png)

The architectures were the same for MLP and GraphSAINT model (same layer dims, activation functions, dropout, learning rate, etc.). 

I expect to see the same performance here. Any idea what the issue might be?

as a side note: I also noticed that when I do an architecture search (with neighbor aggregations), I can get comparable but slightly worse result than an MLP. Given this information with the above, it means that the neighbor information helps the model, but still the GraphSaint "mlp" baseline is very low for some reason.

**Comment by ZimpleX** [comment]
Just to confirm: in your plot, the "GraphNN" curve is for '0-0-0-0' architecture, right? 

You are correct that '0-0-0-0' corresponds to the MLP architecture. I guess the main reason for the performance gap is due to the non iid sampling and the corresponding loss normalization. I assume in your normal MLP training, you would construct the minibatch by taking data points uniformly at random. When training a '0-0-0-0' GraphSAINT with the currently supported subgraph samplers, the minibatch node probability would be non-uniform. To eliminate this factor, we may need to implement a uniform node sampler (different from the "Node" sampler discussed in the paper). Alternatively, you can also compare the performance of full batch training to get rid of the non-uniform minibatch issue. You can set in your `*.yml` config the following: 

```
- end: <num epochs>
  sampler: 'full_batch'
```

Another factor may be in batch-norm. We have three options to perform batch normalization: `bias`, `norm` and `norm-nn`. So `bias` would not perform batch norm; `norm` would normalize features on the per-node basis (which is reasonable due to non-uniform node probability); `norm` would normalize on the per-dimension basis and is implemented by the standard pytorch function using `torch.nn.BatchNorm1d`. Your standard MLP would most likely correspond to either the `bias` or the `norm-nn` setup. 

Also, from my personal experience, the place to perform dropout may also affect hyperparameters. We perform dropout on the input activation (e.g., line 118 of  [layers.py](https://github.com/GraphSAINT/GraphSAINT/blob/master/graphsaint/pytorch_version/layers.py)) where other implementations (e.g., PyG) may perform dropout on the output activation. The location to perform dropout doesn't really affect the model expressivity, but it would result in different choices of the hyperparameter values. It seems that we need to set dropout smaller for our own implementation. 

There are some other minor differences, but they may likely have a less significant impact. e.g., in between the GNN layers and the final classifier layer, we have a L2-normalization layer (see line 116 of [models.py](https://github.com/GraphSAINT/GraphSAINT/blob/master/graphsaint/pytorch_version/models.py)). We initially add this extra normalization by referring to the GraphSAGE official implementation. But later on, I find that this layer may not make a big difference.
